native/rhai_rustler/src/engine.rs

```diff
@@ -35,15 +42,17 @@ fn engine_compile<'a>(
     env: Env<'a>,
     resource: ResourceArc<EngineResource>,
     script: &str,
-) -> Result<Term<'a>, Term<'a>> {
+) -> Result<ResourceArc<ASTResource>, Term<'a>> {
     let engine = resource.engine.try_lock().unwrap();
 
     match engine.compile(script) {
         Ok(result) => {
-            let dynamic = Dynamic::from(result);
-            Ok(from_dynamic(env, dynamic))
+            let ast_resource = ResourceArc::new(ASTResource {
+                ast: Mutex::new(result),
+            });
+            Ok(ast_resource)
         }
-        Err(e) => Err(to_error(env, *e)),
+        Err(_) => Err((atoms::parsing(), "parsing error".to_string().encode(env)).encode(env)),
```

Comment:
ignore what we spoke about on telegram, this works because of rustler magic so it should be ok